### PR TITLE
Enable prioritising tagged definitions.

### LINKED ISSUE #246: Priority for ContextProviders
It is not possible to replace or prioritise the order in which context providers are supplied, meaning that it is not currently possible to consistently remove or add headers to the existing context hash.

### PR DESCRIPTION
**Pull request by antoligy**
Closes #246.

### CODE REVIEW

File: DependencyInjection/Compiler/HashGeneratorPass.php [hunk 1]
```diff
@@ -34,15 +34,30 @@ public function process(ContainerBuilder $container)
...
+
         $providers = array();
-        foreach ($container->findTaggedServiceIds(self::TAG_NAME) as $id => $parameters) {
+        foreach ($prioritisedProviders as $id) {
             $providers[] = new Reference($id);
         }
 
-        if (!count($providers)) {
-            throw new InvalidConfigurationException('No user context providers found. Either tag providers or disable fos_http_cache.user_context');
+        if (!empty($providers)) {
```
Comment (dbu) [inline DependencyInjection/Compiler/HashGeneratorPass.php]: how could providers be empty now? you kept the sanity check with `count` above, so i don't think we need this check.

**Comment by dbu** [comment]
looks good, and i see that this makes sense. i have one comment on the code.

can you please also add a short section to the documentation (in Resources/doc) that explains this feature?

File: DependencyInjection/Compiler/HashGeneratorPass.php (hunk 1 above)
Comment (antoligy) [inline DependencyInjection/Compiler/HashGeneratorPass.php]: This was in response to Scrutinizer's "high" rated issue wherein apparently (???) it was not always possible to fulfil $providers.

**Comment by antoligy** [comment]
Certainly. :smile:

File: DependencyInjection/Compiler/HashGeneratorPass.php (hunk 1 above)
Comment (dbu) [inline DependencyInjection/Compiler/HashGeneratorPass.php]: then that looks like a scrutinizer bug, right?

Comment (antoligy) [inline DependencyInjection/Compiler/HashGeneratorPass.php]: I think so, should I remove it?

Comment (dbu) [inline DependencyInjection/Compiler/HashGeneratorPass.php]: yes, lets remove it.

**Comment by antoligy** [comment]
Just added some docs, and have responded to feedback.

**Comment by antoligy** [comment]
Any further feedback?

**Comment by dbu** [comment]
thanks a lot! i squashed the commits and merged in b24ab0cfdb117d88ff83660b70b03329c2c1b813

**Pull request closed without merging**